### ISSUE TITLE
k'th diagonal version of `diag` missing for `Diagonal`

### ISSUE THREAD

**Issue by jakebolewski** [JuliaLang/julia#9994]
```julia
julia> d = Diagonal(rand(5,5))
5x5 Base.LinAlg.Diagonal{Float64}:
 0.670529  0.0       0.0       0.0       0.0
 0.0       0.755698  0.0       0.0       0.0
 0.0       0.0       0.993278  0.0       0.0
 0.0       0.0       0.0       0.155609  0.0
 0.0       0.0       0.0       0.0       0.224386

julia> diag(d)
5-element Array{Float64,1}:
 0.670529
 0.755698
 0.993278
 0.155609
 0.224386

julia> diag(d,1)
ERROR: MethodError: `similar` has no method matching similar(::Base.LinAlg.Diagonal{Float64}, ::Type{Float64}, ::(Int64,))
```

**Comment by jiahao** [comment]
Again, I would prefer to see this issue resolved as part of subsuming Diagonal, Bidiagonal, SymTridiagonal and Tridiagonal into a general Banded matrix format as proposed in #8240

**Comment by jakebolewski** [comment]
`diag` is also currently not defined for `Symmetric` or `Hermitian`.

This breaks `trace` for these matrix types as well.

**Comment by mbauman** [comment]
Fixed by #10525.
